x/chainlet/keeper/upgrade.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return err
+	}
+
+	// Check the ICA account
+	icaOwner := sdk.AccAddress(address.Module(types.ModuleName)).String()
+	portID, err := icatypes.NewControllerPortID(icaOwner)
+	if err != nil {
+		return err
+	}
+
+	addr, found := k.icaKeeper.GetInterchainAccountAddress(ctx, connectionID, portID)
+	if !found {
+		return fmt.Errorf("missing ICA account for chainlet %s", chainlet.ChainId)
+	}
+
+	// Check ICA channel
+	_, open := k.icaKeeper.GetOpenActiveChannel(ctx, connectionID, portID)
+	if !open {
+		return fmt.Errorf("channel for connection %s and port %s not open", connectionID, portID)
+	}
+
+	// Create a MsgSoftwareUpgrade message
+	clientState, ex := k.ibcKeeper.ClientKeeper.GetClientState(ctx, clientID)
+	if !ex {
+		return fmt.Errorf("client state missing for client ID '%s'", clientID)
+	}
+	upgradeHeight := clientState.GetLatestHeight().GetRevisionHeight() + heightDelta
+	planName, err := upgradePlanName(versionFrom, versionTo)
+	if err != nil {
+		return err
+	}
+	msg := upgradetypes.MsgSoftwareUpgrade{
+		Authority: addr,
+		Plan: upgradetypes.Plan{
+			Name:   planName,
+			Height: int64(upgradeHeight),
+			Info:   "Upgrade created by the provider chain",
+		},
+	}
+
+	// Send the message using ICA
+	data, err := icatypes.SerializeCosmosTx(k.cdc, []proto.Message{&msg}, icatypes.EncodingProtobuf)
+	if err != nil {
+		return err
+	}
+	packetData := icatypes.InterchainAccountPacketData{
+		Type: icatypes.EXECUTE_TX,
+		Data: data,
+	}
+	timeout := ctx.BlockTime().Add(24 * time.Hour).UnixNano() //TODO module param
+	_, err = k.icaKeeper.SendTx(ctx, nil, connectionID, portID, packetData, uint64(timeout))
+	if err != nil {
+		return err
+	}
+
+	// Mark the chainlet as being upgraded
+	err = k.setUpgrading(ctx, chainlet, versionTo, upgradeHeight)
+	if err != nil {
+		return fmt.Errorf("error while updating chainlet: %w", err)
+	}
+
+	return nil
+}
+
+func (k *Keeper) HandleUpgradingChainlets(ctx sdk.Context) error {
+	iterator := prefix.NewStore(ctx.KVStore(k.storeKey), types.UpgradingChainletsKey).Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		chainID := iterator.Key()
+
+		chainlet, err := k.Chainlet(ctx, string(chainID))
+		if err != nil {
+			return err
+		}
+		if chainlet.Upgrade == nil {
+			panic("no upgrade info")
+		}
+
+		clientId, ex := k.providerKeeper.GetConsumerClientId(ctx, chainlet.ChainId)
+		if !ex {
+			//TODO log
+			continue
+		}
+		clientState, ex := k.ibcKeeper.ClientKeeper.GetClientState(ctx, clientId)
+		if !ex {
+			return fmt.Errorf("client state missing for client ID '%s'", clientId)
+		}
+
+		//TODO check revision number
+		height := clientState.GetLatestHeight().GetRevisionHeight()
+		if height > chainlet.Upgrade.Height {
+			// Chain failed to stop before the upgrade height => cancel the upgrade
+			k.cancelUpgrading(ctx, &chainlet)
+			continue
+		}
+		if height < chainlet.Upgrade.Height-1 {
+			continue
+		}
+
+		k.finishUpgrading(ctx, &chainlet)
```

Comment:
This will only trigger if the height is consistently updated every block (which isn't always the case). We should handle the edge case if the relayer skips a client update.